app/controllers/admin_user_controller.rb

```diff
@@ -112,8 +112,9 @@ def clear_profile_photo
   end
 
   def modify_comment_visibility
-    Comment.where(:id => params[:comment_ids]).
-      update_all(:visible => !params[:hide_selected])
+    comments = Comment.where(id: params[:comment_ids])
+    comments.update_all(visible: !params[:hide_selected])
+    comments.find_each(&:reindex_request_events)
```

Comment:
Ideally we shouldn't need to call this in the controller. Issue is we're calling `update_all` I would be tempted to switch to `#update` (or `#toggle`?)

```suggestion
    comments = Comment.where(id: params[:comment_ids], visible: params[:hide_selected])
    comments.find_each(&:toggle, :visible)
```

Would result in `#updated_at` changing and the callbacks being fired too.

Reply:
Also this query should be scoped to the user.

Reply:
> I would be tempted to switch to `#update`…

Yeah, this passes all the specs, so that's a good move. 

```diff
diff --git a/app/controllers/admin_user_controller.rb b/app/controllers/admin_user_controller.rb
index 30e23efcd..f7842fb24 100644
--- a/app/controllers/admin_user_controller.rb
+++ b/app/controllers/admin_user_controller.rb
@@ -113,8 +113,11 @@ def clear_profile_photo

   def modify_comment_visibility
     comments = Comment.where(id: params[:comment_ids])
-    comments.update_all(visible: !params[:hide_selected])
-    comments.find_each(&:reindex_request_events)
+
+    comments.find_each do |comment|
+      comment.update!(visible: !params[:hide_selected])
+    end
+
     redirect_back(fallback_location: admin_users_url)
   end
```

> …or `#toggle`

We're not _toggling_ the visibility – we're setting a specific state for each comment based on the button pressed – so I don't think that would work.

We'd have to add an extra step to filter the comments which aren't already in the state requested by the button press, _then_ we could `toggle!` (n.b. need the bang version to save).

This passes:

```diff
diff --git a/app/controllers/admin_user_controller.rb b/app/controllers/admin_user_controller.rb
index 30e23efcd..fe6750c67 100644
--- a/app/controllers/admin_user_controller.rb
+++ b/app/controllers/admin_user_controller.rb
@@ -112,9 +112,11 @@ def clear_profile_photo
   end

   def modify_comment_visibility
-    comments = Comment.where(id: params[:comment_ids])
-    comments.update_all(visible: !params[:hide_selected])
-    comments.find_each(&:reindex_request_events)
+    comments = Comment.where(id: params[:comment_ids]).
+        where(visible: !params[:hide_selected])
+
+    comments.find_each { |comment| comment.toggle!(:visible) }
+
     redirect_back(fallback_location: admin_users_url)
   end
```

I quite like this, since it's now only updating records which actually need to be updated, rather than updating others already in the correct state as a consequence of the implementation details. It's a little messy as above, but with an extra fix to scope to the user's comments it's much better:

```ruby
  def modify_comment_visibility
    @admin_user.
      comments.
      where(id: params[:comment_ids]).
      where(visible: !params[:hide_selected]).
      find_each { |comment| comment.toggle!(:visible) }

    redirect_back(fallback_location: admin_users_url)
  end
```

Reply:
Doh, as soon as I hit enter I realise I made a mistake. The `toggle!` version above won't work because the of the way we negate only one of the params and that doesn't work in the same way with the `where`.

This is what works:

```ruby
    desired_visibility = params[:hide_selected] ? false : true

    comments =
      Comment.where(id: params[:comment_ids]).
      where(visible: !desired_visibility).
      find_each do |comment|
        comment.toggle!(:visible)
      end
```